dwave/optimization/src/simplex.cpp

```diff
@@ -0,0 +1,493 @@
+// Copyright 2025 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "simplex.hpp"
+
+#include <algorithm>
+
+namespace dwave::optimization {
+
+template <class T>
+void print_vec(std::string name, const std::vector<T>& vec) {
+    std::cout << name << "(" << vec.size() << "): ";
+    for (const T& v : vec) std::cout << v << " ";
+    std::cout << std::endl;
+}
+
+class Matrix {
+ public:
+    Matrix(ssize_t n, ssize_t m) : n_(n), m_(m) { buffer_.resize(n_ * m_); }
+
+    Matrix(std::vector<double>&& data, ssize_t n, ssize_t m)
+            : n_(n), m_(m), buffer_(std::move(data)) {
+        if (n_ * m_ != static_cast<ssize_t>(buffer_.size())) {
+            throw std::invalid_argument("shape does not match provided data");
+        }
+    }
+
+    Matrix(const std::vector<double>& data, ssize_t n, ssize_t m)
+            : Matrix(std::vector<double>(data), n, m) {}
+
+    double& at(ssize_t i, ssize_t j) {
+        if (i < 0) i += n_;
+        if (j < 0) j += m_;
+        assert(i >= 0 && i < n_);
+        assert(j >= 0 && j < m_);
+        return buffer_[i * m_ + j];
+    }
+
+    const double& at(ssize_t i, ssize_t j) const {
+        if (i < 0) i += n_;
+        if (j < 0) j += m_;
+        assert(i >= 0 && i < n_);
+        assert(j >= 0 && j < m_);
+        return buffer_[i * m_ + j];
+    }
+
+    void print(std::string name) const {
+        std::cout << name << ": (" << n_ << ", " << m_ << ")" << std::endl;
+        for (ssize_t i = 0; i < n_; i++) {
+            for (ssize_t j = 0; j < m_; j++) {
+                std::cout << at(i, j) << " ";
+            }
+            std::cout << std::endl;
+        }
+        std::cout << std::endl;
+    }
+
+    ssize_t n() const { return n_; }
+
+    ssize_t m() const { return m_; }
+
+ private:
+    ssize_t n_;
+    ssize_t m_;
+
+    std::vector<double> buffer_;
+};
+
+ssize_t _pivot_col(Matrix& T, double tol, bool bland) {
+    double min_val = std::numeric_limits<double>::infinity();
+    ssize_t col = -1;
+    for (ssize_t j = 0; j < T.m() - 1; j++) {
+        double val = T.at(-1, j);
+        if (val < -tol) {
+            // Bland's rule: return the first index with negative value
+            if (bland) return j;
+
+            if (val < min_val) {
+                min_val = val;
+                col = j;
+            }
+        }
+    }
+    return col;
```

Comment:
Maybe toss an `assert(T.m() * T.n() == 0 || col >= 0)` here? Or, if we don't want to allow empty `Matrix`, then assert that the matrix has nonzero elements.

Applies to `_pivot_row()` as well.

Reply:
These functions return `-1` to mean that no row/column was found. I'm fine to change this to something like using `std::optional` to make it clearer though.

Reply:
No, I think `-1` make sense. But I would definitely mention it in a comment to the intent is clear.